x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentNodeService.java

```diff
@@ -458,17 +577,37 @@ private void handleLoadFailure(TrainedModelDeploymentTask task, Exception ex) {
             "model failed to load; reason [" + ex.getMessage() + "]",
             ActionListener.noop()
         );
-        updateStoredState(
-            task.getModelId(),
-            new RoutingStateAndReason(RoutingState.FAILED, ExceptionsHelper.unwrapCause(ex).getMessage()),
-            ActionListener.wrap(r -> stopTask.run(), e -> stopTask.run())
-        );
+        Optional<RoutingInfo> routingInfo = getRoutingInfo(task.getModelId());
+        if (routingInfo.isEmpty()) {
+            stopTask.run();
+        } else {
+            updateStoredState(
+                task.getModelId(),
+                new RoutingInfo(
+                    routingInfo.get().getCurrentAllocations(),
+                    routingInfo.get().getTargetAllocations(),
+                    RoutingState.FAILED,
+                    ExceptionsHelper.unwrapCause(ex).getMessage()
+                ),
+                ActionListener.wrap(r -> stopTask.run(), e -> stopTask.run())
+            );
+        }
     }
 
     public void failAssignment(TrainedModelDeploymentTask task, String reason) {
+        Optional<RoutingInfo> routingInfo = getRoutingInfo(task.getModelId());
+        if (routingInfo.isEmpty()) {
+            return;
+        }
```

Comment:
relying on clusterService.state here is dangerous. It could be out of sync meaning the state IS there, it just hasn't been fully handled yet. Consequently we get a weird zombie task.

Reply:
I have now pushed a commit that allows partial updates to routing info and eliminates the need to retrieve a routing info based on some state.